renderdoc/driver/vulkan/vk_pixelhistory.cpp

```diff
@@ -438,147 +306,614 @@ struct VulkanPixelHistoryCallback : public VulkanDrawcallCallback
+  }
+
+  void PreDraw(uint32_t eid, VkCommandBuffer cmd)
+  {
+    auto it = m_Events.find(eid);
+    if(it == m_Events.end())
+      return;
+    EventUsage event = it->second;
+    m_PrevState = m_pDriver->GetRenderState();
+
+    // TODO: handle secondary command buffers.
+    // TODO: can't end renderpass if we are not on the last subpass.
+    m_pDriver->GetRenderState().EndRenderPass(cmd);
+    // Get pre-modification values
+    size_t storeOffset = m_EventIndices.size() * sizeof(EventInfo);
+    VkImage depthImage = VK_NULL_HANDLE;
+    VkFormat depthFormat = VK_FORMAT_UNDEFINED;
+    const DrawcallDescription *draw = m_pDriver->GetDrawcall(eid);
+    if(draw && draw->depthOut != ResourceId())
+    {
+      ResourceId resId = m_pDriver->GetResourceManager()->GetLiveID(draw->depthOut);
+      depthImage = m_pDriver->GetResourceManager()->GetCurrentHandle<VkImage>(resId);
+      const VulkanCreationInfo::Image &imginfo = m_pDriver->GetDebugManager()->GetImageInfo(resId);
+      depthFormat = imginfo.format;
+    }
+
+    CopyPixel(m_Image, m_Format, depthImage, depthFormat, cmd, storeOffset);
+
+    VulkanRenderState &pipestate = m_pDriver->GetRenderState();
+    ResourceId prevState = pipestate.graphics.pipeline;
+    ResourceId prevRenderpass = pipestate.renderPass;
+    ResourceId prevFramebuffer = pipestate.GetFramebuffer();
+    rdcarray<ResourceId> prevFBattachments = pipestate.GetFramebufferAttachments();
+    const VulkanCreationInfo::Pipeline &p =
+        m_pDriver->GetRenderState().m_CreationInfo->m_Pipeline[pipestate.graphics.pipeline];
+    uint32_t prevSubpass = pipestate.subpass;
+
+    {
+      PipelineReplacements replacements = GetPipelineReplacements(eid, pipestate.graphics.pipeline);
+
+      if(p.dynamicStates[VkDynamicViewport])
+        for(uint32_t i = 0; i < pipestate.views.size(); i++)
+          ScissorToPixel(pipestate.views[i], pipestate.scissors[i]);
+
+      pipestate.SetFramebuffer(GetResID(m_OffscreenFB));
+      pipestate.renderPass = GetResID(m_RenderPass);
+      pipestate.subpass = 0;
+      pipestate.graphics.pipeline = GetResID(replacements.fixedShaderStencil);
+      ReplayDraw(cmd, (uint32_t)m_OcclusionQueries.size(), eid, true, true);
+
+      m_OcclusionQueries.insert(
+          std::pair<uint32_t, uint32_t>(eid, (uint32_t)m_OcclusionQueries.size()));
+
+      CopyPixelParams params = {};
+      params.multisampled = false;
+      params.srcImage = m_StencilImage;
+      params.srcImageLayout = VK_IMAGE_LAYOUT_DEPTH_STENCIL_ATTACHMENT_OPTIMAL;
+      params.srcImageFormat = VK_FORMAT_D32_SFLOAT_S8_UINT;
+      params.imageOffset.x = int32_t(m_X);
+      params.imageOffset.y = int32_t(m_Y);
+      params.imageOffset.z = 0;
+      params.dstBuffer = m_DstBuffer;
+      params.depthCopy = true;
+      params.stencilOnly = true;
+      // Copy before it is overwritten by original shader stencil.
+      m_pDriver->GetDebugManager()->PixelHistoryCopyPixel(
```

Comment:
It took me a bit of re-reading and refreshing my memory to understand again that we're counting overdraw for discards here in the stencil value (i.e. counting the increments from 0 with a fixed shader that never discards, vs a shader that potentially discards). Could you add some comments here to explain and perhaps rename `fixed_stencil` & `original_stencil` to something to more directly related to overdraw counts and discards?

The D3D11 pixel history I know is really bad for lacking comments and being far too much code all together, so I realise I'm being hypocritical here!

Reply:
Added comments and renamed. Let me know if something is not clear.